### PR TITLE
Add comments for the mask part

### PR DESCRIPTION
**Pull request by YangFei1990**
Comments and documentation for the mask part. Also some for the roi and other parts.

### CODE REVIEW

File: MaskRCNN/model/boxclass_head.py [hunk 1]
```diff
@@ -174,11 +174,13 @@ def f(X):
 @layer_register(log_shape=True)
 def boxclass_2fc_head(feature, fp16=False):
     """
+    Fully connected layer for the class and box branch
+
     Args:
-        feature (any shape):
+        feature: The roi feature, Num_boxes x Num_channels x H x W
```
Comment (armandmcqueen) [inline MaskRCNN/model/boxclass_head.py]: I would call is the feature map instead of `feature`

File: MaskRCNN/model/fpn.py [hunk 2]
```diff
@@ -107,11 +108,11 @@ def fpn_map_rois_to_levels(boxes):
...
     Args:
-        features ([tf.Tensor]): 4 FPN feature level 2-5
-        rcnn_boxes (tf.Tensor): nx4 boxes
-        resolution (int): output spatial resolution
+        features ([tf.Tensor]): 4 FPN feature level P2-5, each with BS X NumChannel X W X H
+        rcnn_boxes (tf.Tensor): t x 5 boxes, t is the number of sampled boxes
+        resolution (int): output spatial resolution, scalar
     Returns:
-        NxC x res x res
+        all_rois: Num_fg_boxes x Num_channels x H x W, H=W=14 if use FPN
```
Comment (armandmcqueen) [inline MaskRCNN/model/fpn.py]: I think 14 is a config param so I would remove `H=W=14 if use FPN`

File: MaskRCNN/model/generalized_rcnn.py [hunk 3]
```diff
@@ -116,6 +117,18 @@ def backbone(self, image):
 
 
     def rpn(self, image, features, inputs, orig_image_dims):
+        """
+        The RPN part of the graph that generate the RPN proposal and losses
+
+        Args:
+            image: BS x Num_channel x H x W
+            features: ([tf.Tensor]): A list of 5 FPN feature level P23456, each with BS x NumChannel x H x W
```
Comment (armandmcqueen) [inline MaskRCNN/model/generalized_rcnn.py]: We are using `Num_channel`/`NumChannel`, `H` and `W` to refer to multiple values. Image is `BS x Num_Channel x H x W`, while features are `BS x NumChannel x fH x fW`. We need to clarify that these variables are not the same and choose a consistent format for `Num_channel` vs `NumChannel`

File: MaskRCNN/model/generalized_rcnn.py [hunk 4]
```diff
@@ -116,6 +117,18 @@ def backbone(self, image):
...
+            inputs: dict, contains all input information
+            orig_image_dims: BS x H x W x Num_channels
```
Comment (armandmcqueen) [inline MaskRCNN/model/generalized_rcnn.py]: I don't think this is right. `orig_image_dims` should be `BSx3`

File: MaskRCNN/model/generalized_rcnn.py [hunk 5]
```diff
@@ -116,6 +117,18 @@ def backbone(self, image):
...
+        Returns:
+            proposal_boxes: top K region proposals, K x 4
```
Comment (armandmcqueen) [inline MaskRCNN/model/generalized_rcnn.py]: Aren't proposal boxes `Kx5`?

File: MaskRCNN/model/generalized_rcnn.py [hunk 6]
```diff
@@ -194,33 +208,47 @@ def rpn(self, image, features, inputs, orig_image_dims):
...
+            image: BS x Num_channel x H x W
+            features: ([tf.Tensor]): A list of 5 FPN feature level P23456, each with BS X NumChannel X W X H
+            proposal_boxes(tf.Tensor): K x 5 boxes
+            targets: list of 'gt_boxes', 'gt_labels', 'gt_masks' from input
+            inputs: dict, contains all input information
+        Returns: all_losses: a list contains box loss and mask loss
+        """
 
-        image_shape2d = inputs['orig_image_dims'][: ,:2]      # BSx2 (h&w)
+        image_shape2d = inputs['orig_image_dims'][: ,:2]      # BS x H x W
```
Comment (armandmcqueen) [inline MaskRCNN/model/generalized_rcnn.py]: I don't think this is right

File: MaskRCNN/model/generalized_rcnn.py [hunk 7]
```diff
@@ -194,33 +208,47 @@ def rpn(self, image, features, inputs, orig_image_dims):
...
         if self.training:
-            input_proposal_boxes = proposal_boxes
-            input_gt_boxes = gt_boxes
-            input_gt_labels = gt_labels
+            input_proposal_boxes = proposal_boxes # K x 5
+            input_gt_boxes = gt_boxes # BS x Num_GT_boxes x 4
+            input_gt_labels = gt_labels # BS x Num_GT_boxes
 
+            # Sample the input_proposal_boxes to make the foreground(fg) box and background(bg) boxes
+            # ratio close to 1:1. proposal_boxes: Num_sampled_boxs x 5, proposal_labels: 1-D Num_sampled_boxes
```
Comment (armandmcqueen) [inline MaskRCNN/model/generalized_rcnn.py]: The ratio is a config param. It does not have to be 1:1

File: MaskRCNN/model/generalized_rcnn.py [hunk 8]
```diff
@@ -194,33 +208,47 @@ def rpn(self, image, features, inputs, orig_image_dims):
...
                     input_proposal_boxes,
                     input_gt_boxes,
                     input_gt_labels,
                     prepadding_gt_counts,
                     batch_size=cfg.TRAIN.BATCH_SIZE_PER_GPU)
 
-        roi_feature_fastrcnn = multilevel_roi_align(features[:4], proposal_boxes, 7)
-
+        # For the box/class branch
+        roi_feature_fastrcnn = multilevel_roi_align(features[:4], proposal_boxes, 7) # Num_sampled_boxes x Num_channels x H x W, H=W=7
```
Comment (armandmcqueen) [inline MaskRCNN/model/generalized_rcnn.py]: We are reusing `H` and `W` to refer to too many different variables

File: MaskRCNN/model/generalized_rcnn.py [hunk 9]
```diff
@@ -194,33 +208,47 @@ def rpn(self, image, features, inputs, orig_image_dims):
...
         fastrcnn_head_func = getattr(boxclass_head, cfg.FPN.BOXCLASS_HEAD_FUNC)
-        head_feature = fastrcnn_head_func('fastrcnn', roi_feature_fastrcnn, fp16=self.fp16)
-
+        # Get head feature from the fully-connected layers
```
Comment (armandmcqueen) [inline MaskRCNN/model/generalized_rcnn.py]: Remove this line. It's not guaranteed to be a fully-connected layer

File: MaskRCNN/model/generalized_rcnn.py [hunk 10]
```diff
@@ -247,49 +276,44 @@ def roi_heads(self, image, features, proposal_boxes, targets, inputs):
 
                 maskrcnn_head_func = getattr(mask_head, cfg.FPN.MRCNN_HEAD_FUNC)
 
+                # For the mask branch. roi_feature_maskrcnn: Num_fg_boxes x Num_channels x H x W, H=W=14
                 roi_feature_maskrcnn = multilevel_roi_align(
                         features[:4], proposal_fg_boxes, 14,
                         name_scope='multilevel_roi_align_mask')
-
+                # Get the mask logits from the conv layers
```
Comment (armandmcqueen) [inline MaskRCNN/model/generalized_rcnn.py]: I think we can remove this line

File: MaskRCNN/model/generalized_rcnn.py [hunk 11]
```diff
@@ -247,49 +276,44 @@ def roi_heads(self, image, features, proposal_boxes, targets, inputs):
...
                 mask_logits = maskrcnn_head_func(
-                        'maskrcnn', roi_feature_maskrcnn, cfg.DATA.NUM_CATEGORY, fp16=self.fp16)   # #fg x #cat x 28 x 28
-
-
-
-
+                        'maskrcnn', roi_feature_maskrcnn, cfg.DATA.NUM_CATEGORY, fp16=self.fp16)   # Num_fg_boxes x num_category x H x W, H=W=28
```
Comment (armandmcqueen) [inline MaskRCNN/model/generalized_rcnn.py]: `H` and `W` again

File: MaskRCNN/model/mask_head.py [hunk 12]
```diff
@@ -11,19 +11,22 @@
 from model.backbone import GroupNorm
 from config import config as cfg
 from utils.mixed_precision import mixed_precision_scope
-
+from performance import print_buildtime_shape, print_runtime_shape
 
 @under_name_scope()
 def maskrcnn_loss(mask_logits, fg_labels, fg_target_masks):
     """
+    Calculate the loss for the mask
```
Comment (armandmcqueen) [inline MaskRCNN/model/mask_head.py]: I don't think we need this line

File: MaskRCNN/model/mask_head.py [hunk 13]
```diff
@@ -11,19 +11,22 @@
...
+
     Args:
-        mask_logits: #fg x #category xhxw
-        fg_labels: #fg, in 1~#class, int64
-        fg_target_masks: #fgxhxw, float32
+        mask_logits: Num_fg_boxes x num_category x H x W, H=W=28
+        fg_labels: 1-D Num_fg_boxes
```
Comment (armandmcqueen) [inline MaskRCNN/model/mask_head.py]: We should keep this info: `in 1~#class, int64`

File: MaskRCNN/model/mask_head.py [hunk 14]
```diff
@@ -11,19 +11,22 @@
...
+        fg_target_masks: Num_fg_boxes x 28 x 28, float32
```
Comment (armandmcqueen) [inline MaskRCNN/model/mask_head.py]: It doesn't have to be 28, right? I think `h` and `w` are dependent on the size of the inputs, not hard-coded values

File: MaskRCNN/model/mask_head.py [hunk 15]
```diff
@@ -55,14 +59,17 @@ def maskrcnn_loss(mask_logits, fg_labels, fg_target_masks):
 @layer_register(log_shape=True)
 def maskrcnn_upXconv_head(feature, num_category, num_convs, norm=None, fp16=False):
     """
+    Get the mask logits from the conv layers
```
Comment (armandmcqueen) [inline MaskRCNN/model/mask_head.py]: I don't think we need this line

File: MaskRCNN/model/fpn.py [hunk 16]
```diff
@@ -75,15 +76,17 @@ def fpn_map_rois_to_levels(boxes):
     Assign boxes to level 2~5.
 
     Args:
-        boxes (nx4):
+        boxes: t x 4, t is the number of sampled boxes
```
Comment (armandmcqueen) [inline MaskRCNN/model/fpn.py]: Is this right? I thought it was tx5

Comment (YangFei1990) [reply MaskRCNN/model/fpn.py]: You are right, it should be tx5
